

# reactive values  ------------------------------------------------------------------
inputFileStats <- reactiveValues(
  stats = NULL
)

# store cell groups that are defined on the fly using the modular 2D plot
groupNames <- reactiveValues(
  namesDF = data.frame()
)

# colors for samples
sampleCols <- reactiveValues(
  colPal = allowedColors
)

# colors for clusters
clusterCols <- reactiveValues(
  colPal = allowedColors
)

# Here, we store projections that are created during the session. These can be selections of cells or other values that
# are not possible to precalculate.
sessionProjections <- reactiveValues(
  prjs = data.frame()
)

# inputDataFunc ----
# loads singleCellExperiment
#   only counts, rowData, and colData are used. Everything else needs to be recomputed
inputDataFunc <- function(inFile) {
  
  stats <- tibble(.rows = length(inFile$datapath))
  stats$names <- inFile$name
  stats$nFeatures <- 0
  stats$nCells <- 0
  
  #
  cat(file = stderr(), paste("reading", inFile$name[1], "\n"))
  fp <- inFile$datapath[1]
  # fp ="scEx.Rds"
  # fp ="../scShinyHubData/patty1A.v2.Rds"
  fpLs <- load(fp)
  scExFound <- FALSE
  for (varName in fpLs) {
    if ("SingleCellExperiment" %in% class(get(varName))) {
      scEx <- get(varName)
      scExFound <- TRUE
    }
  }
  if (!scExFound) {
    return(NULL)
  }
  fdAll <- rowData(scEx)
  pdAll <- colData(scEx)
  exAll <- assays(scEx)[["counts"]]
  stats[1, "nFeatures"] <- nrow(fdAll)
  stats[1, "nCells"] <- nrow(pdAll)
  
  # read multiple files
  if (length(inFile$datapath) > 1) {
    for (fpIdx in 2:length(inFile$datapath)) {
      cat(file = stderr(), paste("reading", inFile$name[fpIdx], "\n"))
      fp <- inFile$datapath[fpIdx]
      fpLs <- load(fp)
      if (!"scEx" %in% fpLs) {
        next()
      }
      fdIdx <- intersect(rownames(fdAll), rownames(rowData(scEx)))
      # if (length(fdIdx) != nrow(fd)) {
      #   cat(file = stderr(), "Houston, there is a problem with the features\n")
      # }
      # fd <- featuredata[fdIdx, ]
      fdAll <- fdAll[fdIdx, ]
      pd1 <- colData(scEx)
      ex1 <- assays(scEx)[["counts"]][fdIdx, ]
      if (sum(rownames(pdAll) %in% rownames(pd1)) > 0) {
        cat(file = stderr(), "Houston, there are cells with the same name\n")
        rownames(pd1) <- paste0(rownames(pd1), "_", fpIdx)
        pdAll <- rbind(pdAll, pd1)
        colnames(ex1) <- rownames(pd1)
      } else {
        pdAll <- rbind(pdAll, pd1)
      }
      stats[fpIdx, "nFeatures"] <- nrow(fd)
      stats[fpIdx, "nCells"] <- nrow(pd1)
      
      exAll <- Matrix::cbind2(exAll[fdIdx, ], ex1)
    }
  }
  exAll <- as(exAll, "dgTMatrix")
  scEx <- SingleCellExperiment(
    assay = list(counts = exAll),
    colData = pdAll,
    rowData = fdAll
  )
  
  cat(stderr(), "Loaded")
  dataTables <- list()
  featuredata <- rowData(scEx)
  # handle different extreme cases for the symbol column (already encountered)
  if (is.factor(featuredata$symbol)) {
    if (levels(featuredata$symbol) == "NA") {
      featuredata$symbol = toupper(rownames(featuredata))
      rowData(scEx) <- featuredata
    }
  }
  if ("symbol" %in% colnames(featuredata)) {
    featuredata$symbol = toupper(featuredata$symbol)
    rowData(scEx) <- featuredata
  }
  
  # dataTables$featuredataOrg <- rowData(scEx)
  dataTables$scEx <- scEx
  dataTables$featuredata <- featuredata
  
  if (is.null(scEx$barcode)) {
    showNotification("scEx doesn't contain barcode column", type = "error")
    return(NULL)
  }
  # some checks
  
  if (sum(is.infinite(assays(scEx)[["counts"]])) > 0) {
    if (!is.null(getDefaultReactiveDomain())) {
      showNotification("scEx contains infinite values",
                       type = "error"
      )
    }
    return(NULL)
  }
  
  if ("sampleNames" %in% names(colData(scEx))) {
    sampNames <- levels(colData(scEx)$sampleNames)
    isolate({
      # sampleCols$colPal <- colorRampPalette(brewer.pal(
      #   n = 6, name =
      #     "PRGn"
      # ))(length(sampNames))
      sampleCols$colPal <- allowedColors[seq_along(sampNames)]
      names(sampleCols$colPal) <- sampNames
    })
  } else {
    showNotification("scEx - colData doesn't contain sampleNames",
                     duration = NULL, type = "error"
    )
  }
  
  if (sum(c("id", "symbol") %in% colnames(rowData(scEx))) < 2) {
    if (!is.null(getDefaultReactiveDomain())) {
      showNotification("scEx - rowData doesn't contain id and/or symbol columns",
                       duration = NULL, type = "error"
      )
    }
  }
  
  if (!sum(c("symbol", "Gene.Biotype", "Description") %in% colnames(featuredata)) == 3) {
    if (!is.null(getDefaultReactiveDomain())) {
      showNotification("featuredata - one of is missing: symbol, Gene.Biotype, Description)",
                       duration = NULL, type = "error"
      )
    }
    if (!"Gene.Biotype" %in% colnames(featuredata)) {
      featuredata$"Gene.Biotype" <- "not given"
    }
    if (!"Description" %in% colnames(featuredata)) {
      featuredata$"Description" <- "not given"
    }
    featuredata$symbol = toupper(featuredata$symbol)
    dataTables$featuredata <- featuredata
  }
  # if (is.null(rowData(dataTables$scEx)$symbol)){
  #
  # }
  
  inputFileStats$stats <- stats
  return(dataTables)
}



# inputData ----
# internal, should not be used by plug-ins
inputData <- reactive({
  start.time <- base::Sys.time()
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "inputDataFunc")
    }
  )
  if (DEBUG) cat(file = stderr(), "DEBUG:inputData\n")
  
  inFile <- input$file1
  if (is.null(inFile)) {
    if (DEBUG) cat(file = stderr(), "inputData: NULL\n")
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/inputData.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file='~/scShinyHubDebug/inputData.RData')
  
  retVal <- inputDataFunc(inFile)
  
  printTimeEnd(start.time, "inputData")
  exportTestValues(inputData = {list(assays(retVal$scEx)[["counts"]], rowData(retVal$scEx), colData(retVal$scEx)) })  
  return(retVal)
})

medianENSGfunc <- function(scEx) {
  geneC <- Matrix::colSums(scEx > 0, na.rm = TRUE)
  return(median(t(geneC)))
}

# medianENSG ----
medianENSG <- reactive({
  start.time <- Sys.time()
  if (DEBUG)cat(file = stderr(), "medianENSG\n")
  
  scEx_log <- scEx_log()
  if (is.null(scEx)) {
    if (DEBUG) {
      cat(file = stderr(), "medianENSG:NULL\n")
    }
    return(0)
  }
  scEx_log <- assays(scEx_log)[[1]]
  if (ncol(scEx_log) <= 1 | nrow(scEx_log) < 1) {
    return(0)
  }
  retVal <- medianENSGfunc(scEx_log)
  printTimeEnd(start.time, "medianENSG")
  exportTestValues(medianENSG = { retVal })
  
  return(retVal)
})

medianUMIfunc <- function(scEx) {
  umiC <- Matrix::colSums(scEx, na.rm = TRUE)
  return(median(t(umiC)))
}

# medianUMI ----
medianUMI <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "medianUMI\n")
  
  scEx <- scEx()
  if (is.null(scEx)) {
    if (DEBUG) {
      cat(file = stderr(), "medianUMI:NULL\n")
    }
    return(0)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/medianUMI.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file='~/scShinyHubDebug/medianUMI.RData')
  scEx <- assays(scEx)[["counts"]]
  retVal <- medianUMIfunc(scEx)
  
  printTimeEnd(start.time, "medianUMI")
  exportTestValues(medianUMI = { retVal })
  return(retVal)
})

# for now we don't have a way to specifically select cells
# we could cluster numbers or the like
# internal, should not be used by plug-ins
useCellsFunc <-
  function(dataTables,
           geneNames,
           rmCells,
           rmPattern,
           keepCells,
           cellKeepOnly) {
    if (DEBUGSAVE) {
      save(file = "~/scShinyHubDebug/useCellsFunc.RData", list = c(ls()))
    }
    # load(file='~/scShinyHubDebug/useCellsFunc.Rdata')
    goodCols <- rep(TRUE, ncol(dataTables$scEx))
    scEx <- assays(dataTables$scEx)[[1]]
    #### start: cells with genes expressed
    # take only cells where these genes are expressed with at least one read
    genesin <- toupper(geneNames)
    genesin <- gsub(" ", "", genesin, fixed = TRUE)
    genesin <- strsplit(genesin, ",")
    genesin <- genesin[[1]]
    
    cellKeep <- toupper(keepCells)
    cellKeep <- gsub(" ", "", cellKeep, fixed = TRUE)
    cellKeep <- strsplit(cellKeep, ",")
    cellKeep <- cellKeep[[1]]
    
    cellKeepOnly <- toupper(cellKeepOnly)
    cellKeepOnly <- gsub(" ", "", cellKeepOnly, fixed = TRUE)
    cellKeepOnly <- strsplit(cellKeepOnly, ",")
    cellKeepOnly <- cellKeepOnly[[1]]
    
    # specifically remove cells
    if (nchar(rmCells) > 0) {
      cellsRM <- toupper(rmCells)
      cellsRM <- gsub(" ", "", cellsRM, fixed = TRUE)
      cellsRM <- strsplit(cellsRM, ",")
      cellsRM <- cellsRM[[1]]
      goodCols[which(toupper(colnames(dataTables$scEx)) %in% cellsRM)] <- FALSE
    }
    
    # remove cells by pattern
    if (nchar(rmPattern) > 0) {
      goodCols[grepl(rmPattern, colnames(dataTables$scEx))] <- FALSE
    }
    
    if (!length(cellKeep) == 0) {
      ids <- which(toupper(colnames(dataTables$scEx)) %in% cellKeep)
      goodCols[ids] <- TRUE
    }
    
    # genes that have to be expressed at least in one of them.
    selCols <- rep(FALSE, length(goodCols))
    if (!length(genesin) == 0) {
      ids <- which(toupper(dataTables$featuredata$symbol) %in% genesin)
      if (length(ids) == 1) {
        selCols <- scEx[ids, ] > 0
      } else if (length(ids) == 0) {
        showNotification(
          "not enough cells, check gene names for min coverage",
          type = "warning",
          duration = NULL
        )
        return(NULL)
      } else {
        selCols <- Matrix::colSums(scEx[ids, ]) > 0
      }
      goodCols <- goodCols & selCols
    }
    
    if (!length(cellKeepOnly) == 0) {
      goodCols[c(1:length(goodCols))] <- FALSE
      ids <- which(toupper(colnames(dataTables$scEx)) %in% cellKeepOnly)
      goodCols[ids] <- TRUE
    }
    
    #### end: cells with genes expressed
    
    return(goodCols)
  }

# useCells ----
# works on cells only
# internal, should not be used by plug-ins
useCells <- reactive({
  start.time <- Sys.time()
  if (DEBUG)cat(file = stderr(), "useCells\n")
  
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "useCells")
    }
  )
  
  dataTables <- inputData()
  geneNames <- input$minExpGenes
  rmCells <- input$cellsFiltersOut
  rmPattern <- input$cellPatternRM
  keepCells <- input$cellKeep
  cellKeepOnly <- input$cellKeepOnly
  # useGenes = isolate(useGenes())
  if (!exists("dataTables") || is.null(dataTables)) {
    if (DEBUG) {
      cat(file = stderr(), "useCells:NULL\n")
    }
    return(NULL)
  }
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("which cells to use", id = "useCells", duration = NULL)
  }
  retVal <- useCellsFunc(
    dataTables,
    geneNames,
    rmCells,
    rmPattern,
    keepCells,
    cellKeepOnly
  )
  
  printTimeEnd(start.time, "useCells")
  exportTestValues(useCells = { retVal })
  return(retVal)
})


useGenesFunc <-
  function(dataTables,
           ipIDs, # regular expression of genes to be removed
           geneListSelection,
           genesKeep,
           geneLists) {
    gList <- geneLists # global variable, assigning it locally ensures that it will be saved
    if (DEBUGSAVE) {
      save(file = "~/scShinyHubDebug/useGenesFunc.Rdata", list = c(ls(), ls(envir = globalenv())))
    }
    # load(file='~/scShinyHubDebug/useGenesFunc.Rdata')
    # regular expression with gene names to be removed
    if (nchar(ipIDs) > 0) {
      keepIDs <- !grepl(ipIDs, dataTables$featuredata$symbol)
    } else {
      keepIDs <- rep(TRUE, nrow(dataTables$scEx))
    }
    genesKeep <- toupper(genesKeep)
    genesKeep <- gsub(" ", "", genesKeep, fixed = TRUE)
    genesKeep <- strsplit(genesKeep, ",")
    genesKeep <- genesKeep[[1]]
    keepGeneIds <- which(dataTables$featuredata$symbol %in% genesKeep)
    
    # dataTables$featuredata$symbol[keepIDs]
    # gene groups to be included
    if (!is.null(geneListSelection)) {
      selectedgeneList <- get_selected(geneListSelection)
      if (length(selectedgeneList) > 0) {
        selGenes <- c()
        for (sIdx in 1:length(selectedgeneList)) {
          print(sIdx)
          att <- attr(selectedgeneList[[sIdx]], "ancestry")
          if (length(att) > 0) {
            selGenes <- c(selGenes, gList[[att]][[selectedgeneList[[sIdx]]]])
          }
        }
        selGenes <- unique(selGenes)
        keepIDs <- (rownames(dataTables$scEx) %in% selGenes) & keepIDs
      }
    }
    
    keepIDs[keepGeneIds] <- TRUE
    return(keepIDs)
  }

# before gene filtering -----
beforeFilterCounts <- reactive({
  start.time <- base::Sys.time()
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "beforeFilterCounts")
    }
  )
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("beforeFilterCounts", id = "beforeFilterCounts", duration = NULL)
  }
  
  dataTables <- inputData()
  ipIDs <- input$selectIds # regular expression of genes to be removed
  
  if (!exists("dataTables") |
      is.null(dataTables) |
      length(dataTables$featuredata$symbol) == 0) {
    if (DEBUG) {
      cat(file = stderr(), "beforeFilterCounts: NULL\n")
    }
    return(NULL)
  }
  
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/beforeFilterCounts.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/beforeFilterCounts.RData")
  
  geneIDs <- NULL
  if (nchar(ipIDs) > 0) {
    geneIDs <- grepl(ipIDs, dataTables$featuredata$symbol)
  }
  if (is.null(geneIDs)) {
    return(rep(0, nrow(dataTables$featuredata)))
  }
  retVal = Matrix::colSums(assays(dataTables$scEx)[["counts"]][geneIDs, ])
  
  printTimeEnd(start.time, "DummyReactive")
  exportTestValues(beforeFilterCounts = { retVal })
  return(retVal)
})

# useGenes ----
# collects information from all places where genes being removed or specified
useGenes <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "useGenes")
    }
  )
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "useGenes\n")
  
  dataTables <- inputData()
  ipIDs <- input$selectIds # regular expression of genes to be removed
  genesKeep <- input$genesKeep
  geneListSelection <- input$geneListSelection
  
  if (!exists("dataTables") |
      is.null(dataTables) |
      length(dataTables$featuredata$symbol) == 0) {
    if (DEBUG) {
      cat(file = stderr(), "useGenes: NULL\n")
    }
    return(NULL)
  }
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("which genes to use", id = "useGenes", duration = NULL)
  }
  
  retVal <- useGenesFunc(dataTables, ipIDs, geneListSelection, genesKeep, geneLists)
  
  printTimeEnd(start.time, "useGenes")
  exportTestValues(useGenes = { retVal })
  return(retVal)
})


# will be called recursively to ensure that nothing changes when cells/genes are changing.
scExFunc <-
  function(scExOrg,
           useCells,
           useGenes,
           minGene,
           minG,
           maxG) {
    if (DEBUG) {
      cat(file = stderr(), "scExFunc\n")
    }
    # if(DEBUG)cat(file=stderr(), paste("col:",ncol(scExOrg),"\n"))
    # if(DEBUG)cat(file=stderr(), paste("l useCells:",length(useCells),"\n"))
    # if(DEBUG)cat(file=stderr(), paste("row:",nrow(scExOrg),"\n"))
    # if(DEBUG)cat(file=stderr(), paste("l useGenes:",length(useGenes),"\n"))
    # scExOrg, useCells, useGenes cannot be NULL
    
    # change names to be hopefully a bit more clear
    changed <- FALSE # trace if something changed
    keepGenes <- useGenes
    keepCells <- useCells
    scEx <- assays(scExOrg)[[1]]
    
    # overall gene expression Min
    if (!is.null(minGene)) {
      selGenes <- Matrix::rowSums(scEx[, keepCells]) >= minGene
      selGenes <- keepGenes & selGenes
      if (!all(selGenes == keepGenes)) {
        keepGenes <- selGenes
        changed <- TRUE
      }
    }
    
    # min reads per cell
    if (!is.null(minG)) {
      selCols <- Matrix::colSums(scEx[keepGenes, ], na.rm = FALSE) > minG
      selCols[is.na(selCols)] <- FALSE
      selCols <- keepCells & selCols
      if (!all(selCols == keepCells)) {
        keepCells <- selCols
        changed <- TRUE
      }
    }
    
    # max reads per cell
    if (!is.null(maxG)) {
      selCols <- Matrix::colSums(scEx[keepGenes, ], na.rm = FALSE) <= maxG
      selCols[is.na(selCols)] <- FALSE
      selCols <- selCols & keepCells
      if (!all(selCols == keepCells)) {
        changed <- TRUE
        keepCells <- selCols
      }
    }
    
    if (sum(keepCells) == 0) {
      showNotification("not enough cells left",
                       type = "warning",
                       duration = NULL
      )
      return(NULL)
    }
    if (sum(keepGenes) == 0) {
      showNotification("not enough genes left",
                       type = "warning",
                       duration = NULL
      )
      return(NULL)
    }
    
    # if something changed, check that it doesn't change again
    scExNew <- scExOrg[keepGenes, keepCells]
    if (changed) {
      scExNew <- scExFunc(scExOrg[keepGenes, keepCells], useCells[keepCells], useGenes[keepGenes], minGene, minG, maxG)
      if (is.null(scExNew)) {
        return(NULL)
      }
    }
    
    pD <- colData(scExNew)
    for (colN in colnames(pD)) {
      if (colN == "barcode") next()
      if (class(pD[, colN]) %in% c("character")) {
        pD[, colN] <- factor(as.character(pD[, colN]))
      }
    }
    colData(scExNew) <- pD
    
    return(scExNew)
  }

# scEx ----
# apply filters that depend on genes & cells
# it is here that useCells and useGenes are combined and applied to select for
scEx <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "scEx")
    }
  )
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "scEx\n")
  
  dataTables <- inputData()
  useCells <- useCells()
  useGenes <- useGenes()
  minGene <- input$minGenesGS # min number of reads per gene
  minG <- input$minGenes # min number of reads per cell
  maxG <- input$maxGenes # max number of reads per cell
  if (!exists("dataTables") |
      is.null(dataTables) | is.null(useGenes) | is.null(useCells)) {
    if (DEBUG) {
      cat(file = stderr(), "scEx: NULL\n")
    }
    return(NULL)
  }
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("scEx", id = "scEx", duration = NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/scEx.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/scEx.RData")
  
  retVal <- scExFunc(
    scExOrg = dataTables$scEx,
    useCells = useCells,
    useGenes = useGenes,
    minGene = minGene,
    minG = minG,
    maxG = maxG
  )
  
  printTimeEnd(start.time, "scEx")
  exportTestValues(scEx = { list(rowData(retVal), colData(retVal)) })
  return(retVal)
})

# rawNormalization ----
rawNormalization <- reactive({
  start.time <- base::Sys.time()
  on.exit(
    if (!is.null(getDefaultReactiveDomain()))
      removeNotification(id = "rawNormalization")
  )
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("rawNormalization", id = "rawNormalization", duration = NULL)
  }
  if (DEBUG) cat(file = stderr(), "rawNormalization\n")
  
  scEx <- scEx()
  
  printTimeEnd(start.time, "rawNormalization")
  exportTestValues(rawNormalization = {str(scEx)})  
  
  return(scEx)
})

# scEx_log ----
scEx_log <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "scEx_log")
    }
  )
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("Normalizing data", id = "scEx_log", duration = NULL)
  }
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "scEx_log\n")
  
  scEx <- scEx()
  normMethod <- input$normalizationRadioButton
  
  if (is.null(scEx)) {
    if (DEBUG) {
      cat(file = stderr(), "scEx_log:NULL\n")
    }
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/scEx_log.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/scEx_log.RData")
  
  scEx_log <- do.call(normMethod, args = list())
  
  printTimeEnd(start.time, "scEx_log")
  exportTestValues(scEx_log = { assays(scEx_log)["logcounts"] })
  return(scEx_log)
})

# scExLogMatrixDisplay ----
# scExLog matrix with symbol as first column
# TODO
# we should probably just rename the rows and then have an option to tableSelectionServer that shows (or not) rownames
scExLogMatrixDisplay <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "scExLogMatrixDisplay")
    }
  )
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "scExLogMatrixDisplay\n")
  
  # dataTables = inputData()
  # useCells = useCells()
  # useGenes = useGenes()
  scEx_log <- scEx_log()
  if (is.null(scEx_log)) {
    if (DEBUG) {
      cat(file = stderr(), "scExLogMatrixDisplay:NULL\n")
    }
    return(NULL)
  }
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("Calculating scExLogmatrix",
                     id = "scExLogMatrixDisplay",
                     duration = NULL
    )
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/scExLogMatrixDisplay.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/scExLogMatrixDisplay.RData")
  
  # TODO
  if (ncol(scEx_log) > 20000) {
    
  }
  retVal <- data.frame(symbol=make.names(rowData(scEx_log)$symbol, unique = TRUE), stringsAsFactors = FALSE)
  retVal <- cbind(retVal,
                          as.matrix(assays(scEx_log)[[1]]))
  rownames(retVal) <- make.names(rowData(scEx_log)$symbol, unique = TRUE)
  
  printTimeEnd(start.time, "scExLogMatrixDisplay")
  return(retVal)
})

pcaFunc <- function(scEx_log) {
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/pcaFunc.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/pcaFunc.RData")
  scaterPCA <- tryCatch({
    # not sure, but this works on another with dgTMatrix
    if (class(assays(scEx_log)[["logcounts"]]) == "dgTMatrix") {
      assays(scEx_log)[["logcounts"]] <- as(assays(scEx_log)[["logcounts"]], "dgCMatrix")
    }
    scater::runPCA(scEx_log,
                   ncomponents = 10, method = "irlba",
                   ntop = 500, exprs_values = "logcounts"
    )
  },
  error = function(e) {
    if (!is.null(getDefaultReactiveDomain())) {
      showNotification(
        "Problem with PCA, probably not enough cells?",
        type = "warning",
        id = "pcawarning",
        duration = NULL
      )
    }
    cat(file = stderr(), "PCA FAILED!!!\n")
    return(NULL)
  }
  )
  if (is.null(scaterPCA)) return(NULL)
  # pca = reducedDim(scaterPCA, "PCA")
  # attr(pca,"percentVar")
  #
  return(list(
    x = SingleCellExperiment::reducedDim(scaterPCA, "PCA"),
    var_pcs = attr(SingleCellExperiment::reducedDim(scaterPCA, "PCA"), "percentVar")
  ))
}

# pca ----
pca <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "pca")
    }
  )
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("pca", id = "pca", duration = NULL)
  }
  start.time <- Sys.time()
  if (!is.null(getDefaultReactiveDomain())) {
    removeNotification(id = "pcawarning")
  }
  if (DEBUG) cat(file = stderr(), "pca\n")
  
  scEx_log <- scEx_log()
  if (is.null(scEx_log)) {
    if (DEBUG) {
      cat(file = stderr(), "pca:NULL\n")
    }
    return(NULL)
  }
  retVal <- pcaFunc(scEx_log)
  
  printTimeEnd(start.time, "pca")
  exportTestValues(pca = { retVal })
  return(retVal)
})


scranCluster <- function(pca, scEx_log, seed, clusterSource,
                         geneSelectionClustering="", minClusterSize=2, clusterMethod="PCA", featureData) {
  set.seed(seed)
  geneid <- geneName2Index(geneSelectionClustering, featureData)
  
  params <- list(
    min.mean = NULL,
    min.size = minClusterSize,
    method = clusterMethod
  )
  if (clusterSource == "PCA") {
    params$x <- t(pca$x)
  } else {
    params$x <- scEx_log
    params$assay.type <- "logcounts"
    if (length(geneid) > 0)
      params$subset.row <- geneid
  }
  
  retVal <- tryCatch({
    do.call("quickCluster", params)
  },
  error = function(e) {
    cat(file = stderr(), paste("\nProblem with clustering", e, "\n\n"))
    return(NULL)
  }#,
  # warning = function(e){
  #   cat(file = stderr(), paste("\nclustering produced Warning:\n",e , "\n"))
  #   return(do.call("quickCluster", params))
  # }
  )
  retVal = data.frame(Barcode = colData(scEx_log)$barcode,
                      Cluster = retVal)
  rownames(retVal) = retVal$Barcode
  return(retVal)
}



# dbCluster ----
dbCluster <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "dbCluster\n")
  
  kNr <- input$kNr
  clustering <- scran_Cluster()
  
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/dbCluster.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/dbCluster.RData")
  
  if (is.null(clustering)) {
    if (DEBUG) {
      cat(file = stderr(), "dbCluster: NULL\n")
    }
    return(NULL)
  }
  
  dbCluster <- clustering$Cluster
  
  #cluster colors
  inCols = list()
  lev <- levels(dbCluster)
  
  inCols <- allowedColors[1:length(lev)]
  names(inCols) = lev
  
  clusterCols$colPal = unlist(inCols)
  
  printTimeEnd(start.time, "dbCluster")
  exportTestValues(dbCluster = { dbCluster })
  return(dbCluster)
})



scran_Cluster <- reactive({
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "scran_Cluster")
    }
  )
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("scran_Cluster", id = "scran_Cluster", duration = NULL)
  }
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "scran_Cluster\n")
  
  pca <- pca()
  scEx_log <- scEx_log()
  seed <- input$seed
  kNr <- input$kNr
  clusterSource <- input$clusterSource
  geneSelectionClustering <- input$geneSelectionClustering
  minClusterSize <- input$minClusterSize
  clusterMethod <- input$clusterMethod
  
  if (is.null(pca) | is.null(scEx_log) | is.na(minClusterSize)) {
    if (DEBUG) {
      cat(file = stderr(), "scran_Cluster:NULL\n")
    }
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/scran_Cluster.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/scran_Cluster.RData")
  
  featureData <- rowData(scEx_log)
  
  if (is.null(seed)) {
    seed <- 1
  }
  retVal <- scranCluster(
    pca, scEx_log, seed, clusterSource,
    geneSelectionClustering, minClusterSize, clusterMethod, 
    featureData
  )
  if (is.null(retVal)) {
    showNotification(
      paste("error: clustering didn't produce a result"),
      type = "error",
      duration = NULL
    )
  }
  
  printTimeEnd(start.time, "scran_Cluster")
  exportTestValues(scran_Cluster = { retVal })
  return(retVal)
})


# projections ----
#' projections
#' each column is of length of number of cells
#' if factor than it is categorical and can be cluster number of sample etc
#' if numeric can be projection
#' projections is a reactive and cannot be used in reports. Reports have to organize
#' themselves as it is done here with tsne.data.
projections <- reactive({
  start.time <- Sys.time()
  if (DEBUG)cat(file = stderr(), "projections\n")
  # scEx is the fundamental variable with the raw data, which is available after loading
  # data. Here we ensure that everything is loaded and all varialbles are set by waiting
  # input data being loaded
  scEx <- scEx()
  pca <- pca()
  prjs <- sessionProjections$prjs
  
  if (!exists("scEx") | is.null(scEx) | !exists("pca") | is.null(pca)) {
    if (DEBUG) {
      cat(file = stderr(), "sampleInfo: NULL\n")
    }
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/projections.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/projections.RData")
  
  projections <- data.frame(pca$x[, c(1, 2, 3)])
  pd <- colData(scEx)
  if (ncol(pd) < 2) {
    cat(file = stderr(), "phenoData for scEx has less than 2 columns\n")
    return(NULL)
  }
  
  withProgress(message = "Performing projections", value = 0, {
    n <- length(projectionFunctions)
    iter <- 1
    for (proj in projectionFunctions) {
      start.time1 <- Sys.time()
      incProgress(1 / n, detail = paste("Creating ", proj[1]))
      if (DEBUG) cat(file = stderr(), paste("calculation projection:  ", proj[1], "\n"))
      assign("tmp", eval(parse(text = paste0(proj[2], "()"))))
      if (DEBUGSAVE) {
        save(file = paste0("~/scShinyHubDebug/projections.", iter, ".RData"), list = c("tmp"))
        iter <- iter + 1
      }
      # load(file="~/scShinyHubDebug/projections.1.RData")
      if (class(tmp) == "data.frame") {
        cn <- make.names(c(colnames(projections), colnames(tmp)))
      } else {
        cn <- make.names(c(colnames(projections), make.names(proj[1])))
      }
      if (length(tmp) == 0) {
        next()
      }
      if (ncol(projections) == 0) {
        # never happening because we set pca first
        projections <- data.frame(tmp = tmp)
      } else {
        if (nrow(projections) == length(tmp)) {
          projections <- cbind(projections, tmp)
        } else if (nrow(projections) == nrow(tmp)) {
          projections <- cbind(projections, tmp)
        } else {
          stop("error: ", proj[1], "didn't produce a result")
        }
        if (DEBUG) {
          end.time <- Sys.time()
          cat(file = stderr(), "===", proj[1], ":done", difftime(end.time, start.time1, units = "min"), "\n")
        }
      }
      
      colnames(projections) <- cn
      observe(proj[2], quoted = TRUE)
    }
  })
  # add a column for gene specific information that will be filled/updated on demand
  projections$UmiCountPerGenes <- 0
  projections$UmiCountPerGenes2 <- 0
  for (pdIdx in colnames(pd)) {
    if (!pdIdx %in% colnames(projections)) {
      projections[, pdIdx] <- pd[, pdIdx]
    }
  }
  
  if (ncol(prjs) > 0 & nrow(prjs) == nrow(projections)) {
    projections <- cbind(projections, prjs)
  }
  
  printTimeEnd(start.time, "projections")
  exportTestValues(projections = { projections })
  return(projections)
})

# initializeGroupNames ----
# TODO shouldn't this be an observer???
initializeGroupNames <- reactive({
  if (DEBUG) cat(file = stderr(), "initializeGroupNames.\n")
  
  scEx <- scEx()
  if (is.null(scEx)) {
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/initializeGroupNames.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/initializeGroupNames.RData")
  isolate({
    df <- data.frame(all = rep(TRUE, ncol(scEx)), none = rep(FALSE, ncol(scEx)))
    rownames(df) <- colnames(scEx)
    cat(file = stderr(), "initializeGroupNames2\n")
    groupNames[["namesDF"]] <- df
    cat(file = stderr(), "initializeGroupNames3\n")
  })
})

# sample --------
sample <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "sample\n")
  
  scEx <- scEx()
  if (is.null(scEx)) {
    if (DEBUG) {
      cat(file = stderr(), "sample: NULL\n")
    }
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/sample.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/sample.RData")
  
  pd <- colData(scEx)
  retVal <- NULL
  for (pdColName in colnames(pd)) {
    if (length(levels(factor(pd[, pdColName]))) < 100) {
      if (is.null(retVal)) {
        retVal <- data.frame(pd[, pdColName])
        colnames(retVal) <- pdColName
      }
      retVal[, pdColName] <- factor(as.character(pd[, pdColName]))
    }
  }
  
  printTimeEnd(start.time, "sample")
  exportTestValues(sample = { retVal })
  return(retVal)
})

# geneCount --------
geneCount <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "geneCount\n")
  
  scEx_log <- scEx_log()
  
  if (is.null(scEx_log)) {
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/geneCount.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/geneCount.RData")
  
  retVal <- Matrix::colSums(assays(scEx_log)[["logcounts"]] > 0)
  
  printTimeEnd(start.time, "geneCount")
  exportTestValues(geneCount = { retVal })
  return(retVal)
})

# beforeFilterPrj ----
beforeFilterPrj <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "umiCount\n")
  
  scEx <- scEx()
  bfc <- beforeFilterCounts()
  
  if (is.null(scEx) | is.null(bfc)) {
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/beforeFilterPrj.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/beforeFilterPrj.RData")
  
  cn <- colnames(scEx)
  retVal <- bfc[cn]
  
  printTimeEnd(start.time, "beforeFilterPrj")
  exportTestValues(beforeFilterPrj = { retVal })
  return(retVal)
})

# umiCount ----
umiCount <- reactive({
  start.time <- Sys.time()
  if (DEBUG) cat(file = stderr(), "umiCount\n")
  
  scEx <- scEx()
  
  if (is.null(scEx)) {
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/umiCount.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/umiCount.RData")
  
  retVal <- Matrix::colSums(assays(scEx)[["counts"]])
  
  printTimeEnd(start.time, "umiCount")
  exportTestValues(umiCount = { retVal })
  return(retVal)
})

# sampleInfoFunc ----
sampleInfoFunc <- function(scEx) {
  # gsub(".*-(.*)", "\\1", scEx$barcode)
  colData(scEx)$sampleNames
}


# sampleInfo -------
# sample information
sampleInfo <- reactive({
  start.time <- Sys.time()
  if (DEBUG)  cat(file = stderr(), "sampleInfo\n")
  
  scEx <- scEx()
  if (!exists("scEx")) {
    if (DEBUG) {
      cat(file = stderr(), "sampleInfo: NULL\n")
    }
    return(NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/sampleInfo.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file="~/scShinyHubDebug/sampleInfo.RData")
  
  retVal <- sampleInfoFunc(scEx)
  
  printTimeEnd(start.time, "sampleInfo")
  exportTestValues(sampleInfo = { retVal })
  return(retVal)
})


# table of input cells with sample information
# TODO: used in tableSeletionServer table; should be divided into function and reactive
inputSample <- reactive({
  start.time <- base::Sys.time()
  on.exit(
    if (!is.null(getDefaultReactiveDomain())) {
      removeNotification(id = "inputSample")
    }
  )
  if (DEBUG) cat(file = stderr(), "inputSample\n")
  
  dataTables <- inputData()
  
  if (is.null(dataTables)) {
    return(NULL)
  }
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("inputSample", id = "inputSample", duration = NULL)
  }
  if (DEBUGSAVE) {
    save(file = "~/scShinyHubDebug/inputSample.RData", list = c(ls(), ls(envir = globalenv())))
  }
  # load(file='~/scShinyHubDebug/inputSample.RData')
  
  # TODO should come from sampleInfo
  sampInf <- gsub(".*-(.*)", "\\1", dataTables$scEx$barcode)
  cellIds <- data.frame(
    cellName = colnames(dataTables$scEx),
    sample = sampInf,
    ngenes = Matrix::colSums(assays(dataTables$scEx)[[1]])
  )
  
  if (DEBUG) {
    cat(file = stderr(), "inputSample: done\n")
  }
  if (dim(cellIds)[1] > 1) {
    return(cellIds)
  } else {
    return(NULL)
  }
})


getMemoryUsed <- reactive({
  require(pryr)
  if (DEBUG) {
    cat(file = stderr(), "getMemoryUsed\n")
  }
  # number of times memory was calculated 
  # not used anymore
  # umu <- updateMemUse$update
  paste(utils:::format.object_size(mem_used(), "auto"))
})

# used in coExpression, subclusterAnalysis, moduleServer, generalQC, DataExploration
# TODO change to scEx_log everywhere and remove
# log2cpm <- reactive({
#   if (DEBUG) {
#     cat(file = stderr(), "log2cpm\n")
#   }
#   scEx_log <- scEx_log()
#   if (is.null(scEx_log)) {
#     if (DEBUG) {
#       cat(file = stderr(), "log2cpm: NULL\n")
#     }
#     return(NULL)
#   }
#   log2cpm <- as.data.frame(as.matrix(assays(scEx_log)[[1]]))
#   
#   return(log2cpm)
# })


reportFunction <- function(tmpPrjFile) {
}

reacativeReport <- reactive({
  start.time <- Sys.time()
  # remove any notification on exit that we don't want
  on.exit(
    if (!is.null(getDefaultReactiveDomain()))
      removeNotification(id = "creating_Report")
  )
  # show in the app that this is running
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("creating Report", id = "creating_Report", duration = NULL)
  }
  
  scEx <- scEx()
  projections <- projections()
  scEx_log <- scEx_log()
  inputNames <- names(input)
  
  if (is.null(scEx)) {
    if (DEBUG) cat(file = stderr(), "output$report:NULL\n")
    return(NULL)
  }
  
  tmpPrjFile <- tempfile(pattern = "file", tmpdir = reportTempDir, fileext = ".RData")
  
  report.env = new.env()
  # translate reactiveValues to lists
  # this way they can be saved
  rectVals = c()
  isolate({
    for (var in c(names(globalenv()),names(parent.env(environment())))) { 
      cat(file = stderr(), paste("var: ", var, '---', class(get(var)),  "\n"))
      if ( var == "reacativeReport") next()
      if (class(get(var))[1] == "reactivevalues"){
        cat(file = stderr(), paste("is reactiveValue: ", var, "\n"))
        rectVals = c(rectVals, var)
        assign(var, reactiveValuesToList(get(var)), envir = report.env)
      } else if (class(get(var))[1] == "reactiveExpr") {
        cat(file = stderr(), paste("is reactiveExpr: ", var, "--", class(get(var)), "\n"))
        rectVals = c(rectVals, var)
        assign(var, eval(parse(text = paste0(var, "()"))), envir = report.env)
        # for modules we have to take care of return values
        # this has to be done manually (for the moment)
        # and is only required for clusterServer
        if (class(report.env[[var]])[1] == "reactivevalues") {
          if(all(c("cluster", "selectedCells") %in% names(report.env[[var]]))) {
            assign(paste0(var,"-cluster"), eval(report.env[[var]][["cluster"]]), envir = report.env)
            tempVar = report.env[[var]][["selectedCells"]]
            assign(paste0(var,"-selectedCells"), eval(parse(text = "tempVar()")), envir = report.env)
          }
        }
      }
    }
    
    assign("input", reactiveValuesToList(get("input")), envir = report.env)
  })
  base::save(file = tmpPrjFile, list = c("reportTempDir","projections", "scEx_log", "scEx", "report.env"))
  userDataEnv <- as.environment(as.list(session$userData, all.names=TRUE))
  # browser()
  if (DEBUGSAVE) save(file = "~/scShinyHubDebug/tempReport.1.RData", list = c("session", "report.env", "file", ls(), ls(envir = globalenv())))
  # load('~/scShinyHubDebug/tempReport.1.RData')
  
  outZipFile = paste0(reportTempDir, "/report.zip")
  
  reactiveFiles <- ""
  
  # fixed files -----------
  tmpFile <- tempfile(pattern = "file", tmpdir = reportTempDir, fileext = ".RData")
  file.copy("geneLists.RData", tmpFile, overwrite = TRUE)
  reactiveFiles <- paste0(reactiveFiles, "load(file=\"", tmpFile, "\")\n", collapse = "\n")
  
  # global variables
  tmpFile <- tempfile(pattern = "file", tmpdir = reportTempDir, fileext = ".RData")
  save(file = tmpFile, list = c("allowedColors"))
  reactiveFiles <- paste0(reactiveFiles, "load(file=\"", tmpFile, "\")\n", collapse = "\n")
  
  # Projections -----
  # projections can contain mannually annotated groups of cells and different normalizations.
  # to reduce complexity we are going to save those in a separate RData file
  
  # return(tmpPrjFile)
  # the reactive.R can hold functions that can be used in the report to reduce the possibility of code replication
  # we copy them to the temp directory and load them in the markdown
  uiFiles <- dir(path = "contributions", pattern = "reactives.R", full.names = TRUE, recursive = TRUE)
  for (fp in c("serverFunctions.R", "reactives.R", uiFiles)) {
    if (DEBUG) cat(file = stderr(), paste("loading: ", fp, "\n"))
    tmpFile <- tempfile(pattern = "file", tmpdir = reportTempDir, fileext = ".R")
    file.copy(fp, tmpFile, overwrite = TRUE)
    reactiveFiles <- paste0(reactiveFiles, "source(\"", tmpFile, "\", local = TRUE)\n", collapse = "\n")
  }
  # otherwise reactive might overwrite projections...
  reactiveFiles <- paste0(reactiveFiles, "load(file=\"", tmpPrjFile, 
                          "\")\nfor (n in names(report.env)) {assign(n,report.env[[n]])}\n", collapse = "\n")
  # encapsulte the load files in an R block
  LoadReactiveFiles <- paste0("\n\n```{r load-reactives, include=FALSE}\n", reactiveFiles, "\n```\n\n")
  
  # handle plugin reports
  # load contribution reports
  # parse all report.Rmd files under contributions to include in application
  uiFiles <- dir(path = "contributions", pattern = "report.Rmd", full.names = TRUE, recursive = TRUE)
  pluginReportsString <- ""
  fpRidx <- 1
  for (fp in uiFiles) {
    if (DEBUG) cat(file = stderr(), paste("loading: ", fp, "\n"))
    tmpFile <- tempfile(pattern = "file", tmpdir = reportTempDir, fileext = ".Rmd")
    file.copy(fp, tmpFile, overwrite = TRUE)
    pluginReportsString <- paste0(
      pluginReportsString,
      "\n\n```{r child-report-", fpRidx, ", child = '", tmpFile, "', eval=TRUE}\n```\n\n"
    )
    fpRidx <- fpRidx + 1
  }
  
  # Copy the report file to a temporary directory before processing it, in
  # case we don't have write permissions to the current working dir (which
  # can happen when deployed).
  tempReport <- file.path(reportTempDir, "report.Rmd")
  
  # tempServerFunctions <- file.path(reportTempDir, "serverFunctions.R")
  # file.copy("serverFunctions.R", tempServerFunctions, overwrite = TRUE)
  
  # create a new list of all parameters that can be passed to the markdown doc.
  
  params <- list(
    # tempServerFunctions = tempServerFunctions,
    # tempprivatePlotFunctions = tempprivatePlotFunctions,
    calledFromShiny = TRUE # this is to notify the markdown that we are running the script from shiny. used for debugging/development
    # save the outputfile name for others to use to save
    # params$outputFile <- file$datapath[1]
  )
  
  
  
  
  
  # for (idx in 1:length(names(input))) {
  #   params[[inputNames[idx]]] <- input[[inputNames[idx]]]
  # }
  params[["reportTempDir"]] <- reportTempDir
  
  file.copy("report.Rmd", tempReport, overwrite = TRUE)
  
  # read the template and replace parameters placeholder with list
  # of paramters
  x <- readLines(tempReport)
  # x <- readLines("report.Rmd")
  paramString <- paste0("  ", names(params), ": NA", collapse = "\n")
  y <- gsub("#__PARAMPLACEHOLDER__", paramString, x)
  y <- gsub("__CHILDREPORTS__", pluginReportsString, y)
  y <- gsub("__LOAD_REACTIVES__", LoadReactiveFiles, y)
  # cat(y, file="tempReport.Rmd", sep="\n")
  cat(y, file = tempReport, sep = "\n")
  
  if (DEBUG) cat(file = stderr(), "output$report:scEx:\n")
  if (DEBUG) cat(file = stderr(), paste("\n", tempReport, "\n"))
  # Knit the document, passing in the `params` list, and eval it in a
  # child of the global environment (this isolates the code in the document
  # from the code in this app)
  if (DEBUG) file.copy(tempReport, "~/scShinyHubDebug/tempReport.Rmd")
  myparams <- params # needed for saving as params is already taken by knitr
  # if (DEBUGSAVE) 
    save(file = "~/scShinyHubDebug/tempReport.RData", list = c("session", "myparams", ls(), "zippedReportFiles"))
  # load(file = '~/scShinyHubDebug/tempReport.RData')
  cat(file = stderr(), paste("workdir: ", getwd()))
  require(callr)
  # if (DEBUGSAVE)
    file.copy(tempReport, "~/scShinyHubDebug/tmpReport.Rmd", overwrite = TRUE)
  
  tempReport = "~/scShinyHubDebug/tmpReport.Rmd"
  file.copy("contributions/coE_coExpression/report.Rmd",
            '/var/folders/_h/vtcnd09n2jdby90zkb6wyd740000gp/T//Rtmp5uuY0f/file92087e386a11.Rmd', overwrite = TRUE)
  r(function(input, output_file, params, envir) 
    rmarkdown::render(input = input, output_file = output_file,
                                                                  params = params, envir = envir), 
    args = list(input = tempReport,
                output_file = "report.html",
                params = params,
                envir = new.env())
  )
  # file.copy(from = "contributions/sCA_subClusterAnalysis/report.Rmd",
  #           to = "/var/folders/_h/vtcnd09n2jdby90zkb6wyd740000gp/T//Rtmph1SRTE/file69aa37a47206.Rmd", overwrite = TRUE)
  # rmarkdown::render(input = tempReport, output_file = "report.html",
  #                   params = params, envir = new.env())
  
  tDir <- paste0(reportTempDir, "/")
  base::save(file = paste0(reportTempDir, "/sessionData.RData"), list = c(ls(), ls(envir = globalenv())))
  write.csv(as.matrix(assays(scEx_log)[[1]]), file = paste0(reportTempDir, "/normalizedCounts.csv"))
  base::save(file = paste0(reportTempDir, "/inputUsed.Rds"), list = c("scEx", "projections"))
  zippedReportFiles <- c(paste0(tDir, zippedReportFiles))
  zip(outZipFile, zippedReportFiles, flags = "-9Xj")
  if (!is.null(getDefaultReactiveDomain())) {
    showNotification("Report creation is done", id = "reportDone", duration = 10, type = "message")
  }
  if (DEBUG) {
    end.time <- Sys.time()
    cat(file = stderr(), "===Report:done", difftime(end.time, start.time, units = "min"), "\n")
  }
  return(outZipFile)
})
