---
title: "Single Cell Analysis Report"
author: "scShinyHub"
date: "3/8/2018"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 7
    fig_width: 9
    number_sections: yes
    toc: yes
params:
#__PARAMPLACEHOLDER__
---

```{r checkDEBUG}
if(!exists("DEBUG")){
  DEBUG=TRUE
  # DEBUG=FALSE
}
```

```{r loadData}
if(exists("params")){
  cat(file=stderr(), paste("params:", params$calledFromShiny,"\n"))
  cat(file=stderr(), paste("params exists:", "calledFromShiny" %in% names(params) ,"\n"))
}else{
  # rm(list = ls())
  require(shiny)
  require(shinyTree)
  require(shinyBS)
  require(plotly)
  require(shinythemes)
  require(ggplot2)
  require(DT)
  require(pheatmap)
  require(threejs)
  require(sm)
  require(RColorBrewer)
  require(mclust)
  require(reshape)
  require(cellrangerRkit)
  require(SCORPIUS)
  require(ggplot2)
  require(knitr)
  require(kableExtra)
  require(shinyWidgets)
  require(scater)
  
  source("serverFunctions.R")
  source("privatePlotFunctions.R")
  source("reactives.R", local = TRUE)
  uiFiles = dir(path = "contributions", pattern = "reactives.R", full.names = TRUE, recursive = TRUE)
  for(fp in uiFiles){
    if(DEBUG)cat(file=stderr(), paste("loading: ", fp, "\n"))
    source(fp, local = TRUE)
  }
  load("~/Desktop/report.RData")
  LOCALEXECUTION = TRUE # to know that we are debugging.
}

```


__LOAD_REACTIVES__

# parameters

```{r setup, include=FALSE}
input = params
print(str(input$file1))
if(file.exists(params$file1$datapath)){
  load(params$file1$datapath)
}
print(ls())
```



```{r check Variables}
if(is.null(input$cluster)) {
  input$cluster = '0'
  print("Warning: setting cluster to 0")
}
if(is.null(input$cluster5)) {
  input$cluster5 = '0'
  print("Warning: setting cluster5 to 0")
}
if(is.null(input$clusters)) {
  input$clusters = '0'
  print("Warning: setting clusters to 0")
}
if(is.null(input$clusters1)) {
  input$clusters1 = '0'
  print("Warning: setting clusters1 to All")
}
if(is.null(input$clusters2)) {
  input$clusters2 = '0'
  print("Warning: setting clusters2 to 0")
}
if(is.null(input$clusters3)) {
  input$clusters3 = '0'
  print("Warning: setting clusters4 to 0")
}
if(is.null(input$clusters4)) {
  input$clusters4 = 'All'
  print("Warning: setting clusters4 to All")
}

```


```{r inputData}
if(! exists("dataTables")) # needed when working with ~/Desktop/reports.Data
  dataTables = inputDataFunc(input$file1)
```

```{r medianENSG}
medianENSG = medianENSGfunc(log2cpm)
```


```{r medianUMI}
medianUMI = medianUMIfunc(log2cpm)
```

```{r useCells}
useCells = useCellsFunc(dataTables)
```

```{r useGenes}
ipIDs = input$selectIds
geneListSelection = input$geneListSelection
minGene <- input$minGenesGS

useGenes = useGenesFunc(dataTables, useCells, ipIDs, geneListSelection, minGene)
```


```{r featureDataReact}

featureDataReact = dataTables$featuredata[useGenes, ]
featureData = dataTables$featuredata[useGenes, ]
```

```{r gbm}
if(DEBUG)cat(file=stderr(), "gbm\n")
gbm = dataTables$gbm[useGenes, useCells]
if(DEBUG)cat(file=stderr(), "gbm:DONE\n")

```

```{r gbm_log}
if(DEBUG)cat(file=stderr(), "gbm_log\n")
gbm_log = dataTables$gbm_log[useGenes, useCells]
if(DEBUG)cat(file=stderr(), "gbm_log:Done\n")

```


```{r log2cpm, eval=FALSE}
if(DEBUG)cat(file=stderr(), "log2cpm\n")
if(DEBUG)cat(file=stderr(), "log2cpm:Done\n")

log2cpm = dataTables$log2cpm[useGenes, useCells]
```

```{r pca, eval=FALSE}
if(DEBUG)cat(file=stderr(), "pca\n")

pca = pcaFunc(gbm_log)
```

```{r tsne, eval=FALSE}
if(DEBUG)cat(file=stderr(), "tsne\n")
seed=1
set.seed(seed = seed)
tsne = run_tsne(pca, dims = 3, perplexity = 30, theta = 0.5)
if(DEBUG)cat(file=stderr(), "tsne: done\n")
```

```{r kmClustering, eval=FALSE}
if(DEBUG)cat(file=stderr(), "kmClustering\n")
clustering=list()

kNr = 10
for(kNr in 2:10) {
  set.seed(seed = seed)
  km = run_kmeans_clustering(pca, k=kNr)
  clustering[[paste0("kmeans_",kNr,"_clusters")]] = data.frame("Barcode" = rownames(data.frame(km$cluster)), "Cluster" = km$cluster)
}

kmClustering = clustering
if(DEBUG)cat(file=stderr(), "kmClustering:done\n")

```

```{r tsne.data, eval=FALSE}
if(DEBUG)cat(file=stderr(), "tsne.data\n")

tsne.data = data.frame(tsne$Y)
colnames(tsne.data) = c("V1", "V2", "V3")
tsne.data$dbCluster = clustering$kmeans_10_clusters$Cluster-1
rownames(tsne.data) = clustering$kmeans_10_clusters$Barcode
if(DEBUG)cat(file=stderr(), "tsne.data: done\n")
```

```{r prioritized_genes, eval=FALSE}
set.seed(seed = seed)
prioritized_genes =     prioritize_top_genes(gbm, tsne.data$dbCluster, "sseq",
                                             logscale = FALSE, 
                                             min_mean=0.5, 
                                             p_cutoff=0.05,
                                             order_by='pvalue')


```

```{r dge, eval=FALSE}
if(DEBUG)cat(file=stderr(), "dge\n")
if(!is.null(input$db1) & ! is.null(input$db2)){
  subsetData <- subset(tsne.data, dbCluster == input$clusters1)
  cells.1 <- rownames(brushedPoints(subsetData, input$db1))
  cells.2 <- rownames(brushedPoints(subsetData, input$db2))
  
  subsetExpression <- log2cpm[, union(cells.1, cells.2)]
  
  genes.use <- rownames(subsetExpression)
  data.1 = apply(subsetExpression[genes.use, cells.1], 1, expMean)
  data.2 = apply(subsetExpression[genes.use, cells.2], 1, expMean)
  total.diff = (data.1 - data.2)
  
  genes.diff = names(which(abs(total.diff) > .2))
  genes.use = ainb(genes.use, genes.diff)
  
  toReturn <-
    DiffExpTest(subsetExpression, cells.1, cells.2, genes.use = genes.use)
  toReturn[, "avg_diff"] = total.diff[rownames(toReturn)]
  toReturn$Associated.Gene.Name <-
    featureData[rownames(toReturn), 'Associated.Gene.Name']
  selectedDge <- toReturn
  cat(stderr(), rownames(toReturn)[1:5])
  
  dge <-     toReturn
  if(DEBUG)cat(file=stderr(), "dge:done11\n")
}else{
  dge=NA
}
```


```{r selectedDge, eval=FALSE}
# selectedDge <- reactiveValues()

```

```{r summaryStats, eval=FALSE}


line1<-paste('No. of cells:', dim(log2cpm)[2],sep='\t')
line2<-paste('Median UMIs:', medianUMI,sep='\t')
line3<-paste('Median Genes:', medianENSG,sep='\t')
line5<-paste('No. of reads:',  dim(log2cpm)[1],sep='\t')
HTML(
  paste0("Summary statistics of this dataset:", '<br/>','<br/>',
         line1, '<br/>', 
         line2, '<br/>',
         line3, '<br/>',
         line5
  )
)

```





```{r selectedGenesTable, eval=FALSE}
gbmMat = as.matrix(exprs(dataTables$gbm))
fd = dataTables$featuredata
dt = fd[useGenes,c("Associated.Gene.Name", "Gene.Biotype", "Description")]
dt$rowSums = rowSums(gbmMat[useGenes,useCells])
dt$rowSamples = rowSums(gbmMat[useGenes,useCells]>0)
DT::datatable(dt)

```

```{r removedGenesTable, eval=FALSE}
gbmMat = as.matrix(exprs(dataTables$gbm))
fd = dataTables$featuredata
dt = fd[useGenes,c("Associated.Gene.Name", "Gene.Biotype", "Description")]
dt$rowSums = rowSums(gbmMat[useGenes,useCells])
dt$rowSamples = rowSums(gbmMat[useGenes,useCells]>0)
DT::datatable(dt)

```


```{r clusterPlot, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$clusterPlot\n")
geneid <- rownames(featureData[which(featureData$Associated.Gene.Name ==
                                       toupper(input$gene_id)), ])[1]

expression <- log2cpm[geneid, ]


tsneData <- cbind(tsne.data, t(expression))
names(tsneData)[names(tsneData) == geneid] <- 'values'


if(is.null(input$cluster)) input$cluster = 0
subsetData <- subset(tsneData, dbCluster == input$cluster)
p1 <-
  ggplot(subsetData,
         aes_string(x = input$dimension_x, y = input$dimension_y)) +
  geom_point(aes_string(size = 2, color = 'values')) +
  geom_point(shape = 1,
             size = 4,
             colour = "black") +
  theme_bw() +
  theme(
    axis.text.x = element_text(
      angle = 90,
      size = 12,
      vjust = 0.5
    ),
    axis.text.y = element_text(size = 12),
    strip.text.x = element_text(size = 16),
    strip.text.y = element_text(size = 14),
    axis.title.x = element_text(face = "bold", size = 16),
    axis.title.y = element_text(face = "bold", size = 16),
    legend.position = "none"
  ) +
  ggtitle(paste(toupper(input$gene_id), input$cluster, sep = '-Cluster')) +
  scale_colour_gradient2(low = 'grey50', high = "red")
p1

```


```{r heatmap, eval=FALSE}
genesin <- input$heatmap_geneids
genesin <- toupper(genesin)
genesin <- gsub(" ", "", genesin, fixed = TRUE)
genesin <- strsplit(genesin, ',')

map <- rownames(featureData[which(featureData$Associated.Gene.Name %in% genesin[[1]]), ])
cat(file = stderr(), length(map))

expression <- log2cpm[map, ]

if(is.na(sum(expression)) == TRUE){
  print('Gene symbol incorrect or genes not expressed')
}else{
  
  tsne.data <- tsne.data[order(tsne.data$dbCluster), ]
  
  expression <- expression[, rownames(tsne.data)]
  expression <- expression[complete.cases(expression), ]
  
  annotation <- data.frame(factor(tsne.data$dbCluster))
  rownames(annotation) <- colnames(expression)
  colnames(annotation) <- c('Cluster')
  
  h <-
    pheatmap(
      as.matrix(expression),
      cluster_rows = TRUE,
      cluster_cols = FALSE,
      scale = 'row',
      fontsize_row = 10,
      labels_col = colnames(expression),
      labels_row = featureData[rownames(expression), 'Associated.Gene.Name'],
      show_rownames = TRUE,
      annotation_col = annotation,
      show_colnames = FALSE,
      annotation_legend = TRUE,
      breaks = seq(-6, 6, by = .12),
      colorRampPalette(rev(brewer.pal(
        n = 6, name =
          "RdBu"
      )))(100)
      
    )
  print(h)
}
```

```{r clusterPlot2, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$clusterPlot2\n")
geneid <- rownames(featureData[which(featureData$Associated.Gene.Name ==
                                       toupper(input$gene_id_sch)), ])[1]

expression <- log2cpm[geneid, ]

validate(need(is.na(sum(expression)) != TRUE, ''))

tsneData <- cbind(tsne.data, t(expression))
names(tsneData)[names(tsneData) == geneid] <- 'values'

# if(DEBUG)cat(file=stderr(), paste("output$dge_plot1:---",input$clusters2,"---\n"))
if(!is.null(input$dimension_x2) & !is.null(input$dimension_y2)){
  subsetData <- subset(tsneData, dbCluster == input$clusters2)
  p1 <-
    ggplot(subsetData,
           aes_string(x = input$dimension_x2, y = input$dimension_y2)) +
    geom_point(aes_string(size = 2, color = 'values')) +
    geom_point(shape = 1,
               size = 4,
               colour = "black") +
    theme_bw() +
    theme(
      axis.text.x = element_text(
        angle = 90,
        size = 12,
        vjust = 0.5
      ),
      axis.text.y = element_text(size = 10),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 14),
      axis.title.x = element_text(face = "bold", size = 16),
      axis.title.y = element_text(face = "bold", size = 16),
      legend.position = "none"
    ) +
    ggtitle(paste(toupper(input$gene_id_sch), input$clusters2, sep =
                    '-Cluster')) +
    scale_colour_gradient2(low = 'grey50', high = "red")
  print(p1)
}
```

```{r selectedHeatmap, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$selectedHeatmap\n")

genesin <- input$heatmap_geneids2
genesin <- toupper(genesin)
genesin <- strsplit(genesin, ',')
if(!is.null(input$scb1)){
  subsetData <-
    subset(tsne.data, dbCluster == input$clusters2)
  cells.1 <- rownames(brushedPoints(df=subsetData, 
                                    brush=input$scb1,
                                    xvar = input$scb1$mapping$x,
                                    yvar = input$scb1$mapping$y,
                                    panelvar1 = 
  ))
  map <- rownames(featureData[which(featureData$Associated.Gene.Name %in% genesin[[1]]), ])
  
  expression <- log2cpm[map, cells.1]
  
  expression <- expression[complete.cases(expression), ]
  cat(file = stderr(), rownames(expression))
  mColor <- max(expression)
  
  if( is.na(sum(expression)) == TRUE){
    print('Gene symbol incorrect or genes not expressed')
  }else{
    
    
    h <-
      pheatmap(
        as.matrix(expression),
        cluster_rows = TRUE,
        cluster_cols = TRUE,
        scale = 'row',
        fontsize_row = 10,
        labels_col = colnames(expression),
        labels_row = featureData[rownames(expression), 'Associated.Gene.Name'],
        show_rownames = TRUE,
        show_colnames = FALSE,
        breaks = seq(-6, 6, by = .12),
        colorRampPalette(rev(brewer.pal(
          n = 6, name =
            "RdBu"
        )))(100)
        
      )
    print(h)
  }
}
```


```{r plotCoExpression, eval=FALSE}

if(DEBUG)cat(file=stderr(), "output$plotCoExpression\n")

genesin <- input$mclustids
genesin <- toupper(genesin)
genesin <- strsplit(genesin, ',')

subsetData <-
  subset(tsne.data, dbCluster == input$clusters3)
cells.1 <- rownames(subsetData)


map <- rownames(featureData[which(featureData$Associated.Gene.Name %in% genesin[[1]]), ])

expression <- log2cpm[map, ]

if(is.na(sum(expression)) == TRUE){
  print ('Gene symbol incorrect or genes not expressed')
}else{
  
  bin <- expression
  bin[] <- 0
  
  for (i in 1:nrow(expression))
  {
    x <- Mclust(expression[i, ], G = 2)
    bin[i, ] <- x$classification
  }
  bin <- bin - 1
  allexprs <- apply(bin, 2, sum)
  plotexprs <- allexprs
  plotexprs[] <- 0
  plotexprs[allexprs >= length(rownames(bin))] <- 1
  positiveCells <- allexprs >= length(rownames(bin))
  positiveCellsAll <- plotexprs
  mergeExprs <- plotexprs[rownames(subsetData)]
  
  subsetData$CoExpression <- mergeExprs
  
  p1 <-
    ggplot(subsetData,
           aes_string(x = input$dimension_x3, y = input$dimension_y3)) +
    geom_point(aes_string(size = 2, color = 'CoExpression')) +
    geom_point(shape = 1,
               size = 4,
               colour = "black") +
    theme_bw() +
    theme(
      axis.text.x = element_text(
        angle = 90,
        size = 12,
        vjust = 0.5
      ),
      axis.text.y = element_text(size = 12),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 14),
      axis.title.x = element_text(face = "bold", size = 16),
      axis.title.y = element_text(face = "bold", size = 16),
      legend.position = "none"
    ) +
    #ggtitle(paste(toupper(input$gene_id),input$cluster,sep='-Cluster'))+
    scale_colour_gradient2(low = 'grey50', high = "red")
  print(p1)
}

```


```{r onOffTable, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$onOffTable\n")

merge <- tsne.data
# if(DEBUG)cat(file=stderr(), paste("positiveCellsAll:---",positiveCellsAll,"---\n"))

merge$CoExpression <- positiveCellsAll
df <-
  as.data.frame(table(merge[, c('dbCluster', 'CoExpression')]))
dfOut <- cast(df, dbCluster ~ CoExpression)
colnames(dfOut) <- c("Cluster", 'OFF', 'ON')
rownames(dfOut) <- dfOut$Cluster
dfOut['Sum', ] <- c('', sum(dfOut$OFF), sum(dfOut$ON))
DT::datatable(dfOut)

```


```{r dge_plot1, eval=FALSE}

if(!is.null(input$dimension_x1) & !is.null(input$dimension_y1)){
  subsetData <- subset(tsne.data, dbCluster == input$clusters1)
  p1 <-
    ggplot(subsetData,
           aes_string(x = input$dimension_x1, y = input$dimension_y1)) +
    geom_point() +
    geom_point(shape = 1,
               size = 4,
               color = "black") +
    theme_bw() +
    theme(
      axis.text.x = element_text(
        angle = 90,
        size = 12,
        vjust = 0.5
      ),
      axis.text.y = element_text(size = 12),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 14),
      axis.title.x = element_text(face = "bold", size = 16),
      axis.title.y = element_text(face = "bold", size = 16),
      legend.position = "none"
    ) +
    ggtitle(input$clusters1)
  print(p1)
}
```


```{r dge_plot2, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$dge_plot2\n")
if(!is.null(input$dimension_x1) & !is.null(input$dimension_y1)){
  subsetData <- subset(tsne.data, dbCluster == input$clusters1)
  p1 <-
    ggplot(subsetData,
           aes_string(x = input$dimension_x1, y = input$dimension_y1)) +
    geom_point() +
    geom_point(shape = 1,
               size = 4,
               color = "black") +
    theme_bw() +
    theme(
      axis.text.x = element_text(
        angle = 90,
        size = 12,
        vjust = 0.5
      ),
      axis.text.y = element_text(size = 12),
      strip.text.x = element_text(size = 16),
      strip.text.y = element_text(size = 14),
      axis.title.x = element_text(face = "bold", size = 16),
      axis.title.y = element_text(face = "bold", size = 16),
      legend.position = "none"
    ) +
    ggtitle(input$clusters1)
  print(p1)
}
```


```{r dgeTable, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$dge\n")

if(!is.na(dge)){
  top.genes <- dge
  top.genes$Associated.Gene.Name <-
    featureData[rownames(top.genes), 'Associated.Gene.Name']
  if (dim(top.genes)[1] > 1) {
    DT::datatable(top.genes,
                  options = list(
                    orderClasses = TRUE,
                    lengthMenu = c(10, 30, 50),
                    pageLength = 10
                  ))
  }
}
```


```{r crHeat_plot1, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$crHeat_plot1\n")

example_K <- 10 
example_Cols <- rev(brewer.pal(10,"Set3")) # customize plotting colors

cells_to_plot <- order_cell_by_clusters(gbm, tsne.data$dbCluster)

example_col = example_Cols[1:example_K]
p = gbm_pheatmap(gbm=log_gene_bc_matrix(gbm), 
                 genes_to_plot=prioritized_genes, 
                 cells_to_plot=cells_to_plot,
                 n_genes=10, 
                 colour=example_col,
                 limits = c(-3, 3))

if(DEBUG)cat(file=stderr(), "output$crHeat_plot1:done\n")
p

```


```{r crPrioGenes, eval=FALSE}
if(DEBUG)cat(file=stderr(), "output$crPrioGenes\n")

dt = prioritized_genes[[input$cluster5]]
DT::datatable(dt)

```


```{r externalContributions, results='asis', echo=FALSE}
# here we process the full set of report files and below print them.
# there will be no table of content
# it is only useful when clicking the "Knit" button in RStudio
# individual chunks can be run locally for debugging/development purposes. See Dummy package or other packages for examples.
options(knitr.duplicate.label = 'allow')
if(LOCALEXECUTION){
  uiFiles = dir(path = "contributions", pattern = "report.Rmd", full.names = TRUE, recursive = TRUE)
  fpRidx = 1
  retReport = NULL
  retString = ""
  for(fp in uiFiles){
    if(DEBUG)cat(file=stderr(), paste("loading: ", fp, "\n"))
    
     fpStr=readLines(fp)
    retReport = c(retReport, knit2html(text=fpStr, quiet = TRUE, fragment.only = TRUE))
    fpRidx = fpRidx + 1
  }
  paste(retReport, collapse = "\n")
}

```

```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```

__CHILDREPORTS__


```{r}
print(ls())
if(!LOCALEXECUTION){
  save(file = "~/Desktop/report.RData", list=ls())
}
sessionInfo()
```

